### PR TITLE
Ngram block for beam search, blocks all repeating ngrams up to n

### PR DESCRIPTION
**Pull request by uralik**
Simply counts ngrams of current hypothesis and blocks whatever was repeated. Solves local repetition issue quite well. Be default we do not block anything. 

PEP styles should be OK

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -897,6 +897,11 @@ def __init__(self, beam_size, min_length=3, padding_token=0, bos_token=1,
         self.eos_top_ts = None
         self.n_best_counter = 0
         self.min_n_best = min_n_best
+        self.block_ngram = block_ngram
+
+    @staticmethod
+    def find_ngrams(input_list, n):
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: wouldn't mind a proper docstring here, but (y)

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -924,6 +929,20 @@ def advance(self, softmax_probs):
...
+                if self.block_ngram > 0:
+                    current_ngrams = []
+                    for ng in range(self.block_ngram):
+                        ngrams = Beam.find_ngrams(current_hypo, ng)
+                        if len(ngrams) > 0:
+                            current_ngrams.extend(ngrams)
+                    counted_ngrams = Counter(current_ngrams)
+                    if any(v > 1 for k, v in counted_ngrams.items()):
+                        # block this hypothesis hard
+                        beam_scores[i] = -1e20
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: use `parlai.core.utils.NEAR_INF`

File: parlai/core/torch_agent.py (hunk 1 above)
Comment (uralik) [reply parlai/core/torch_agent.py]: added

File: parlai/core/torch_agent.py (hunk 2 above)
Comment (uralik) [reply parlai/core/torch_agent.py]: fixed

**Pull request merged**